parser/parser_test.go

```diff
@@ -0,0 +1,173 @@
+	"github.com/madlambda/spells/assert"
+)
+
+type testcase struct {
+	input       string
+	expected    ast.Node
+	expectedErr error
+}
+
+var E = fmt.Errorf
+
+func TestParserNumbers(t *testing.T) {
+	for _, tc := range []testcase{
+		{
+			input:    "1",
+			expected: ast.NewIntNumber(1),
+		},
+		{
+			input:    "1234567890",
+			expected: ast.NewIntNumber(1234567890),
+		},
+		{
+			input:       "1a",
+			expectedErr: E("tests.js:1:0: invalid token: 1a"),
+		},
+		{
+			input:    "0x0",
+			expected: ast.NewIntNumber(0),
+		},
+		{
+			input:    "0x1234567890abcdef",
+			expected: ast.NewIntNumber(0x1234567890abcdef),
+		},
+		{
+			input:    "0xff",
+			expected: ast.NewIntNumber(0xff),
+		},
+		{
+			input:    ".1",
+			expected: ast.NewNumber(0.1),
+		},
+		{
+			input:    ".0000",
+			expected: ast.NewNumber(0.0),
+		},
+		{
+			input:    "1234",
+			expected: ast.NewIntNumber(1234),
+		},
+		{
+			input:    "0.12345",
+			expected: ast.NewNumber(0.12345),
+		},
+		{
+			input:       "0.a",
+			expectedErr: E("tests.js:1:0: invalid token: 0.a"),
+		},
+		{
+			input:       "12.13.",
+			expectedErr: E("tests.js:1:0: invalid token: 12.13."),
+		},
+		{
+			input:    "1.0e10",
+			expected: ast.NewNumber(1.0e10),
+		},
+		{
+			input:    "1e10",
+			expected: ast.NewNumber(1e10),
+		},
+		{
+			input:    ".1e10",
+			expected: ast.NewNumber(.1e10),
+		},
+		{
+			input:    "1e-10",
+			expected: ast.NewNumber(1e-10),
+		},
+		{
+			input: "-1",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(1),
+			),
+		},
+		{
+			input: "-1234",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(1234),
+			),
+		},
+		{
+			input: "-0x0",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(0),
+			),
+		},
+		{
+			input: "-0xff",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(255),
+			),
+		},
+		{
+			input: "-.0",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(0),
+			),
+		},
+		{
+			input: "-.0e1",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(0),
+			),
+		},
+		{
+			input:       "-12.13.",
+			expectedErr: E("tests.js:1:0: invalid token: 12.13."),
+		},
+		{
+			input: "-1e-10",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewNumber(1.0e-10),
+			),
+		},
+		{
+			input: "-+0",
+			expected: ast.NewUnaryExpr(
+				token.Minus, ast.NewUnaryExpr(
+					token.Plus, ast.NewNumber(0),
+				),
+			),
+		},
+		{
+			input: "-+-+0",
+			expected: ast.NewUnaryExpr(token.Minus,
+				ast.NewUnaryExpr(token.Plus,
+					ast.NewUnaryExpr(token.Minus,
+						ast.NewUnaryExpr(token.Plus,
+							ast.NewNumber(0))))),
+		},
+	} {
+		tree, err := parser.Parse("tests.js", tc.input)
+		assert.EqualErrs(t, tc.expectedErr, err, "parser err")
+
+		if err != nil {
+			continue
+		}
+
+		nodes := tree.Nodes
+		if len(nodes) != 1 {
+			t.Fatalf("number tests must be isolated: %v", nodes)
+		}
+
+		got := nodes[0]
+		if got.Type() != tc.expected.Type() {
```

Comment:
Won't Equal already test if the type is the same ?

Reply:
yes, but it's a bit opaque. Some times test fail with message "Numbers differs: 2.0 != 2.0" ... but it's because they were parsed into different types but Equal only says they're not equal...